Added stubs for aiofiles package

erictraut opened the pull request.

JelleZijlstra commented on third_party/3/aiofiles/base.pyi:
```diff
@@ -0,0 +1,39 @@
+
+class _ContextManager(Generic[_V_co, _T_co, _T_contra], Coroutine[_V_co, _T_co, _T_contra]):
+    def __init__(self, __coro: Coroutine[_V_co, _T_co, _T_contra]) -> None: ...
+    def send(self, __value: _T_contra) -> _T_co: ...
```
Mypy is unhappy about a number of incompatible overrides here. I think you should be able to remove most methods from this class, since they're the same as in the Coroutine base class.

erictraut commented:
@JelleZijlstra I could use some advice on this one. Mypy is complaining about overlapping overloads that return different types. The overloads are for the function `open` and are almost identical to those defined in `builtins.pyi` (just a few additional parameters added). I don't see any problems with them. Is this perhaps a bug in mypy's overload analysis?

gvanrossum commented:
It seems that for builtins.open(), the return type of the fallback overload, BinaryIO, is a supertype of the return type of the other two overloads. But in your stub, AsyncFileIO is just a sibling of the other return types. Mypy sees that there is some overlap in the parameters possible: in particular the Literal[-1, 1] for buffering in the earlier overloads overlaps with the int used in the final overload, and insists that the return type varies correspondingly.

erictraut commented:
Thanks, that explains why mypy is emitting the error, but is that error appropriate? I can't think of a reason why this restriction is needed, but maybe I'm missing something. I don't think I can accurately describe the interface of this library with this restriction in place.

gvanrossum commented:
I think it's explained here: https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-variants (search for "overlapping"). This explains that mypy doesn't straightforwardly choose the first match -- if there are overlapping matches it insists that the return type does not depend on which of the overlapping alternatives was used. The examples there have always convinced me that this is a good rule.

The problem with the code here is that the overlap between `Literal[-1, 1]` and `int` is hard to avoid -- we don't have an intersection or exclusion type. Translating this to the example in the docs:
```
file: str = ...
buffering: int = 1
x = aiofiles.threadpool.open(file, "rb", buffering)
```
What is the type of `x`? At runtime it will be `AsyncBufferedReader`, but mypy will pick `AsyncFileIO`, which is not a superclass of the former. To make mypy totally happy you'd have to make it return `AsyncBase`, but that doesn't have any of the useful methods. You could try adding `# type: ignore` directives to the `def` lines mentioned in the errors, but that would cause `x` in the above example to be inferred incorrectly. Maybe a Protocol representing the shared interface between these overloads could be created?

erictraut commented:
@hauntsaninja I think I've addressed all of the issues now. The CI tests appear to have gotten stuck. Perhaps you or someone else with write access could requeue the tests?

gvanrossum commented:
Closing and reopening to re-trigger the tests.
